asmcomp/cmmgen.ml

```diff
@@ -2421,6 +2441,23 @@ and transl_prim_2 env p arg1 arg2 dbg =
   | Pintcomp cmp ->
       tag_int(Cop(Ccmpi(transl_int_comparison cmp),
                   [transl env arg1; transl env arg2], dbg)) dbg
+  | Pcompare_ints ->
+      (* Compare directly on tagged ints *)
+      let a1 = transl env arg1 in
+      let a2 = transl env arg2 in
+      mk_compare_ints dbg a1 a2
+  | Pcompare_bints bi ->
+      let a1 = transl_unbox_int dbg env bi arg1 in
+      let a2 = transl_unbox_int dbg env bi arg2 in
+      mk_compare_ints dbg a1 a2
+  | Pcompare_floats ->
+      let a1 = transl_unbox_float dbg env arg1 in
+      let a2 = transl_unbox_float dbg env arg2 in
+      let op1 = Cop(Ccmpf(CFgt), [a1; a2], dbg) in
+      let op2 = Cop(Ccmpf(CFlt), [a1; a2], dbg) in
+      let op3 = Cop(Ccmpf(CFeq), [a1; a1], dbg) in
+      let op4 = Cop(Ccmpf(CFeq), [a2; a2], dbg) in
```

Comment:
This looks correct, but it took me a while to puzzle it out and a comment would help!

(My understanding is: if either operand is `nan`, then `op1 = op2 = 0` and `op3 - op4` orders `nan` before non-`nan` values. If neither operand is `nan`, then `op3 - op4` is `0` (both are `1`), so `op1 - op2` gives the answer)